pkg/compactor/executor.go

```diff
@@ -283,8 +351,160 @@ func (e *schedulerExecutor) updateJobStatus(ctx context.Context, key *compactors
+		if hasNonZeroULIDs(compactedBlockIDs) {
+			compactor.metrics.groupCompactions.Inc()
+		}
+		level.Info(userLogger).Log("msg", "compaction job completed", "tenant", userID, "compacted_blocks", len(compactedBlockIDs))
+		return compactorschedulerpb.COMPLETE, nil
+	}
+
+	// At this point the compaction has failed. Track the failure.
+	compactor.metrics.groupCompactionRunsFailed.Inc()
+
+	if ok, issue347Err := isIssue347Error(err); ok {
+		level.Warn(userLogger).Log("msg", "detected issue347 error during compaction", "block", issue347Err.id, "err", err)
+		repairErr := repairIssue347(ctx, userLogger, userBucket, syncer.metrics.blocksMarkedForDeletion, issue347Err)
+		if repairErr == nil {
+			level.Info(userLogger).Log("msg", "successfully repaired issue347 block, job can be re-planned", "block", issue347Err.id)
+			compactor.metrics.groupCompactionRunsCompleted.Inc()
+			return compactorschedulerpb.COMPLETE, nil
+		}
+		level.Error(userLogger).Log("msg", "failed to repair issue347 block, abandoning job", "block", issue347Err.id, "err", repairErr)
+		return compactorschedulerpb.ABANDON, errors.Wrap(repairErr, "failed to repair issue347 block")
+	}
+
+	if ok, outOfOrderChunksErr := IsOutOfOrderChunkError(err); ok && compactor.skipUnhealthyBlocks {
+		level.Warn(userLogger).Log("msg", "detected out-of-order chunks error during compaction", "block", outOfOrderChunksErr.id, "err", err)
+		markErr := e.retryable.WithContext(ctx).Run(func() error {
+			return block.MarkForNoCompact(
+				ctx,
+				userLogger,
+				userBucket,
+				outOfOrderChunksErr.id,
+				block.OutOfOrderChunksNoCompactReason,
+				"OutofOrderChunk: marking block with out-of-order series/chunks as no compact to unblock compaction",
+				compactor.metrics.blocksMarkedForNoCompact.WithLabelValues(block.OutOfOrderChunksNoCompactReason),
+			)
+		})
+
+		if markErr == nil {
+			compactor.metrics.groupCompactionRunsCompleted.Inc()
+			level.Info(userLogger).Log("msg", "marked block for no-compact due to out-of-order chunks, job can be re-planned", "block", outOfOrderChunksErr.id)
+			return compactorschedulerpb.COMPLETE, nil
+		}
+		level.Warn(userLogger).Log("msg", "failed to mark block for no-compact after retries", "block", outOfOrderChunksErr.id, "err", markErr)
+	}
+
+	if ok, criticalErr := IsCriticalError(err); ok && compactor.skipUnhealthyBlocks {
+		level.Warn(userLogger).Log("msg", "detected critical error during compaction", "block", criticalErr.id, "err", err)
+		markErr := e.retryable.WithContext(ctx).Run(func() error {
+			return block.MarkForNoCompact(
+				ctx,
+				userLogger,
+				userBucket,
+				criticalErr.id,
+				block.CriticalNoCompactReason,
+				"UnhealthyBlock: marking unhealthy block as no compact to unblock compaction",
+				compactor.metrics.blocksMarkedForNoCompact.WithLabelValues(block.CriticalNoCompactReason),
+			)
+		})
+
+		if markErr == nil {
+			compactor.metrics.groupCompactionRunsCompleted.Inc()
```

Comment:
This was a critical error. It seems incorrect to mark this as complete (after being marked as failed on `448`) in metrics and to the scheduler.